### PR TITLE
Fetch albums in folders for iCloud Photos

### LINKED ISSUE #288: Get Photo-Album within Folder
How can I access Photo Albums, that are with ICloud-Folders? Tried 'Folder/Albumname' and 'Folder\Albumname'. Did not work.

### PR DESCRIPTION
**Pull request by noizwaves**
## Proposed change

This PR enhances the PhotosService to traverse sub-folders when it retrieves albums. The initial implementation only shallowly iterated over items in the root folder. The code has been changed to first determine if an item is actually a folder, and if so to query iCloud for any albums inside of it.

The ancestry or "lineage" of an album is also tracked in a new field `PhotoAlbum.lineage`; the names of each parent directory are included in this list. The definition of `PhotoAlbum.__unicode__` has been enhanced to include this information in a path-esque way. The aim is to better represent the actual "path" within an iCloud Library.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New service (thank you!)
- [x] New feature (which adds functionality to an existing service)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Documentation or code sample

## Example of code:

```python
for name, album in api.photos.albums.items():
    print(name, album.lineage, str(album))
```

Additionally, I used a local build of this package in [this script](https://github.com/noizwaves/icloud-photo-album-downloader/blob/master/download-albums.py) to download the details of all of my albums. My library consists of albums at the root level, albums in a single folder, and albums nested within multiple levels of folders.

## Additional information

- This PR fixes or closes issue: fixes #288
- This PR is related to issue: 

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated to README

### CODE REVIEW

File: pyicloud/services/photos.py
Comment (ixs) [inline pyicloud/services/photos.py:182]: I see what you did here. And I really do not like it. 😂
I'm just not a fan of "magical numbers" being used throughout the code as they are hard to memorize.

How about something like this:

folder_types = {"album":0, "folder": 3}

if folder["fields"]["albumType"]["value"] == folder_types["album"]:
   ...

That makes it easy to read and the logic becomes clearer.

File: pyicloud/services/photos.py
Comment (ixs) [inline pyicloud/services/photos.py:519]: I do not believe, this is a good logic here. In iCloud Photos, a folder name can contain a "/".

So either you need to escape the / character in foldernames (and possible also Album and Filenames) or you skip the whole thing.
Considering that none of the current services (ubuiquity, files and photos) support Unix-Styled foldernames/paths but instead do it via chaining of list items I'm thinking that maybe dropping this kind of feature completely is not the worst idea.

But let#s see what @Quentame thinks...

File: pyicloud/services/photos.py
Comment (ixs) [inline pyicloud/services/photos.py]: I like this. I might need to steal something like that for my Drive changes...

**Comment by noizwaves** [comment]
Thanks for the feedback @ixs! I've just fixed up the failing CI checks (formatting issues), so I think we're technically mergeable now. Do you still want a review from @Quentame?

**Comment by ixs** [comment]
@noizwaves I saw the changes. I liked them. Thanks for that. But yes, @Quentame has to review. I got no commit rights on this repo. And he's on vacation for a few more days. So patience. 😁

File: pyicloud/services/photos.py
Comment (Quentame) [inline pyicloud/services/photos.py:236]:
```suggestion
            if sub_folder["fields"]["albumType"]["value"] == self.ALBUM_TYPES["album"]:
                self._construct_album(folder_hierarchy, sub_folder)
            elif sub_folder["fields"]["albumType"]["value"] == self.ALBUM_TYPES["folder"]:
```

File: pyicloud/services/photos.py
Comment (Quentame) [inline pyicloud/services/photos.py:230]: Why checking the value here ?

Should we fetch sub folders if the parent is deleted ?

File: pyicloud/services/photos.py
Comment (Quentame) [inline pyicloud/services/photos.py:129]: If you know other types, you can fill it even if we don't use it now.

**Review by Quentame — CHANGES_REQUESTED** [review]
A little change about the "magical numbers" 😄 

Then, it seems nice to me.

File: pyicloud/services/photos.py
Comment (noizwaves) [reply pyicloud/services/photos.py:129]: Unfortunately I don't know what the other types are.

File: pyicloud/services/photos.py
Comment (noizwaves) [reply pyicloud/services/photos.py:230]: Great point; this logic exists already at https://github.com/picklepete/pyicloud/pull/297/files/2e4d3eb62ffc3b36bf3d30e60ac526f48cd366c4#diff-dbc5e8af14fd471636e609e5ecd1fe3bR178. I'll take this out and see what happens.

Comment (noizwaves) [reply pyicloud/services/photos.py:230]: I'm inclined to leave this logic in, in other words, I'm included to not fetch sub folders / albums for deleted parent folders.

In the iCloud GUI, if you delete a folder that contains sub folders, it will delete both the folder and the sub-folder. I think it may be confusing/inconsistent if some deleted folders are traversed when looking for albums, as the results generated by pyicloud will be different than what is visible in the UI. This also remains aligned with the existing code where deleted albums were not returned.

I could see it being valuable to allow the user to choose how deleted folders/abums should be treated when fetching albums in a future PR that extends this functionality.

Comment (Quentame) [reply pyicloud/services/photos.py:230]: OK but I think we should add a quick comment to explain how it is.

Comment (Quentame) [reply pyicloud/services/photos.py:230]: Also we can extract the code here in a little private function as, as you pointed out, the logic exists already at L178

Comment (noizwaves) [reply pyicloud/services/photos.py]: @Quentame I've implemented both of your suggestions, thank you for making them. Sorry for the slow delay in implementing them!

I think all of the feedback has been addressed and this can be merged. Please let me know if you have other feedback.

**Comment by kpiwko** [comment]
Hello @Quentame, what is the the status on this PR?

**Comment by Roelio81** [comment]
> Hello @Quentame, what is the the status on this PR?

@noizwaves and @Quentame, is there still something that needs to be done for this PR or do you need help on anything?
I'm looking forward to getting this PR completed!

**Comment by mm21** [comment]
Hi, can this be pulled please? This is a crucial feature for a project I'm working on. Thanks!

**Pull request closed without merging**
